### PR TITLE
feat: More lifecycle improvements

### PR DESCRIPTION
**Pull request by chirino**
- Integration state decoupled from deployment state, APIs defined to manage deployment.
- Replaced IntegrationDeploymentSpec with a plain Integration

to publish integration - PUT to /integrations/{id}/deployments
to undeploy integration - POST to /integrations/{id}/deployments/{version}/targetState

/integrations/{id}/history removed, replaced with a /integrations/{id}/overview

### CODE REVIEW

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java [hunk 1]
```diff
@@ -36,388 +38,434 @@
...
 
-
     @JsonIgnore
-    public boolean isInactive() {
+    default boolean isInactive() {
         return getCurrentState() == IntegrationDeploymentState.Undeployed || getCurrentState() == IntegrationDeploymentState.Inactive
-            ? true
-            : false;
+                ? true
+                : false;
```
Comment (zregvart) [inline app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: do we need the ternary here?

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java [hunk 2]
```diff
@@ -36,388 +38,434 @@
...
+//            this.createdDate = createdDate.orElse(new Date());
+//            return this;
+//        }
+//
+//        public Builder lastUpdated(Optional<Date> lastUpdated) {
+//            this.lastUpdated = lastUpdated.orElse(new Date());
+//            return this;
+//        }
+//
+//    }
```
Comment (zregvart) [inline app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: let's not do this, git versions source code and we have no need for commented out large blocks of code, remove it rather

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 3]
```diff
@@ -99,23 +99,76 @@ public Kind resourceKind() {
...
+    @SuppressFBWarnings("URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD")
+    public static class IntegrationOverview {
+
+        public int version;
+        public Optional<String> id;
+        public String name;
+        public SortedSet<String> tags;
+        public Optional<String> description;
+        public boolean draft;
+        public List<DeploymentOverview> deployments;
```
Comment (zregvart) [inline app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: perhaps this should be `final` and the Set and list should be unmodifiable?

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 4]
```diff
@@ -99,23 +99,76 @@ public Kind resourceKind() {
...
+
+    @SuppressFBWarnings({"URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD", "URF_UNREAD_FIELD"})
+    public static class DeploymentOverview {
+
+        private Optional<String> id;
+        private int version;
+        private IntegrationDeploymentState currentState;
+        private IntegrationDeploymentState targetState;
+        private Date createdDate;
+        private int integrationVersion;
```
Comment (zregvart) [inline app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: Would prefer the fields to be `final`, and `createdDate` to be `ZonedDateTime`

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 5]
```diff
@@ -145,82 +196,63 @@ public Integration create(@Context SecurityContext sec, @ConvertGroup(from = Def
         Integration encryptedIntegration = encryptionSupport.encrypt(integration);
 
         Integration updatedIntegration = new Integration.Builder()
+            .version(1)
```
Comment (zregvart) [inline app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: also the default value, perhaps not needed?

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java [hunk 6]
```diff
@@ -243,38 +245,26 @@ private void addToExport(Map<String, ModelData<?>> export, ModelData<?> model) {
             switch (model.getKind()) {
                 case Integration: {
                     Integration integration = (Integration) model.getData();
-                    integration = new Integration.Builder()
+                    Integration.Builder builder = new Integration.Builder()
                         .createFrom(integration)
-                        .desiredStatus(IntegrationDeploymentState.Draft)
-                        .build();
+                        .deleted(false)
+                        .version(1)
```
Comment (zregvart) [inline app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: `1` is default, perhaps not necessary

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java [hunk 7]
```diff
@@ -18,16 +18,18 @@
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+
 import io.syndesis.core.IndexedProperty;
 import io.syndesis.core.SuppressFBWarnings;
+import io.syndesis.core.json.StringTrimmingConverter;
```
Comment (zregvart) [inline app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: unused

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java [hunk 8]
```diff
@@ -18,16 +18,18 @@
...
 import io.syndesis.model.Kind;
 import io.syndesis.model.WithId;
 import io.syndesis.model.WithKind;
-import io.syndesis.model.WithName;
 import io.syndesis.model.action.ConnectorAction;
+import io.syndesis.model.validation.UniqueProperty;
+import io.syndesis.model.validation.UniquenessRequired;
```
Comment (zregvart) [inline app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: these two are unused

File: app/rest/controllers/src/main/java/io/syndesis/controllers/integration/online/ActivateHandler.java [hunk 9]
```diff
@@ -126,11 +126,11 @@ public StateUpdate execute(IntegrationDeployment integrationDeploymentDefinition
     private DeploymentData createDeploymentData(Integration integration, IntegrationDeployment integrationDeployment) {
         Properties applicationProperties = IntegrationSupport.buildApplicationProperties(integrationDeployment, dataManager, encryptionComponent);
 
-        String username = integration.getUserId().orElseThrow(() -> new IllegalStateException("Couldn't find the user of the integration"));
+        String username = integrationDeployment.getUserId().orElseThrow(() -> new IllegalStateException("Couldn't find the user of the integration"));
 
         return DeploymentData.builder()
             .addLabel(OpenShiftService.INTEGRATION_ID_LABEL, Labels.sanitize(integrationDeployment.getIntegrationId().orElseThrow(() -> new IllegalStateException("IntegrationDeployment should have an integrationId"))))
-            .addLabel(OpenShiftService.DEPLOYMENT_ID_LABEL, integrationDeployment.getVersion().orElse(0).toString())
+            .addLabel(OpenShiftService.DEPLOYMENT_ID_LABEL, ""+integrationDeployment.getVersion())
```
Comment (zregvart) [inline app/rest/controllers/src/main/java/io/syndesis/controllers/integration/online/ActivateHandler.java]: integrationDeployment.getVersion().toString()

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 10]
```diff
@@ -42,9 +43,14 @@
 import javax.ws.rs.core.SecurityContext;
 import javax.ws.rs.core.UriInfo;
 
+import org.springframework.security.core.context.SecurityContextHolder;
```
Comment (zregvart) [inline app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: unused

File: app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java [hunk 11]
```diff
@@ -48,14 +48,12 @@
     public void deserializeModelDataTest() throws IOException {
 
         Integration integrationIn = new Integration.Builder()
-                    .desiredStatus(IntegrationDeploymentState.Active)
                     .tags(new TreeSet<>(Arrays.asList("tag1", "tag2")))
                     .createdDate(new Date())
                     .build();
         String integrationJson = mapper.writeValueAsString(integrationIn);
         System.out.println(integrationJson);
```
Comment (zregvart) [inline app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java]: perhaps we can do without System.out.println

File: app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java [hunk 12]
```diff
@@ -48,14 +48,12 @@
...
         Integration integrationOut = mapper.readValue(integrationJson, Integration.class);
-        Assert.assertEquals(integrationIn.getDesiredStatus(), integrationOut.getDesiredStatus());
```
Comment (zregvart) [inline app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java]: not sure why the assert was removed

File: app/core/model/src/main/java/io/syndesis/model/integration/Integration.java [hunk 13]
```diff
@@ -50,19 +50,25 @@ default Kind getKind() {
     //TODO: This is actually the deployment version and not the deployment. At some point we need to rename it.
     Optional<Integer> getDeploymentId();
 
-    Optional<String> getUserId();
```
[the remark this answers is not in the document]
Comment (chirino) [reply app/core/model/src/main/java/io/syndesis/model/integration/Integration.java]: It moved to deployments.

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java [hunk 14]
```diff
@@ -36,388 +38,434 @@
...
+    default IntegrationDeploymentState getTargetState() {
+        return IntegrationDeploymentState.Active;
     }
 
-    @Override
-    @JsonProperty("name")
-    public String getName() {
-        return name;
+    @Value.Default
+    default Date getLastUpdated() {
```
[the remark this answers is not in the document]
Comment (chirino) [reply app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: I would personally prefer to switch to keep it simple and switch to a long.

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java (hunk 1 above)
Comment (chirino) [reply app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: Was existing code.  I’d have review closer.

File: app/rest/controllers/src/main/java/io/syndesis/controllers/integration/online/ActivateHandler.java (hunk 9 above)
Comment (chirino) [reply app/rest/controllers/src/main/java/io/syndesis/controllers/integration/online/ActivateHandler.java]: Won’t work

File: app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java (hunk 12 above)
Comment (chirino) [reply app/rest/dao/src/test/java/io/syndesis/dao/ReadApiClientDataTest.java]: Field does not exist anymore.

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java (hunk 1 above)
Comment (chirino) [reply app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: right. dont need it.

File: app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java (hunk 2 above)
Comment (chirino) [reply app/core/model/src/main/java/io/syndesis/model/integration/IntegrationDeployment.java]: removing..

**Comment by chirino** [comment]
some fields have changed: lastUpdated => updatedAt and createdDate => createdAt
They are now holding a long ms since epoch

File: app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java (hunk 6 above)
Comment (chirino) [reply app/rest/rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: right

**Comment by gashcrumb** [comment]
> to publish integration - PUT to /integrations/{id}/deployments

So this is only for existing integrations, right?  Currently for publishing new integrations it's a `PUT` to `/integrations` with `desiredStatus` set to `Active`.

> to undeploy integration - POST to /integrations/{id}/deployments/{version}/targetState

Just to clarify is this just to change the state of the active deployment?  I kinda feel like state management should just be at the integration level, i.e. `POST` to `/integrations/{id}/targetState`?  This would allow the UI's current start/stop controls to remain a bit agnostic about deployments.

**Comment by chirino** [comment]
@gashcrumb 

> So this is only for existing integrations, right? Currently for publishing new integrations it's a PUT to /integrations with desiredStatus set to Active.

In the case of a new integration, PUT to `/integrations`, then follow it up with a PUT to `/integrations/{id}/deployments`

> Just to clarify is this just to change the state of the active deployment? I kinda feel like state management should just be at the integration level, i.e. POST to /integrations/{id}/targetState? This would allow the UI's current start/stop controls to remain a bit agnostic about deployments.

So this kinda muddies the water a bit.  I want clean up  Integration model so that the only time it gets updated is when a user makes changes to the integration steps.

**Comment by gashcrumb** [comment]
> So this kinda muddies the water a bit. I want clean up Integration model so that the only time it gets updated is when a user makes changes to the integration steps.

Makes sense, can I suggest maybe we have a top-level `/integrationDeployments` (or `/deployments`) that the UI could use for the list/dashboard list?  Those two views are concerned with the runtime state of the deployment.  If we take the status etc out of the Integration object, then the lists will have to make a second call to fetch the deployment so the status can be shown.

I guess though with `/deployments` we wouldn't get back drafts :-/

**Comment by chirino** [comment]
squashed and rebased.. but there seems to be another merge needed.

**Comment by chirino** [comment]
@gashcrumb rebased.

**Comment by chirino** [comment]
Codacy looks hung.

**Pull request merged**